tests/utils/test_augmentation.py

```diff
@@ -0,0 +1,251 @@
+        ),
+        param(  # 2 invariant cols, 2 dependent ones, 2 additional ones
+            {
+                "Slot1": ["s1", "s2"],
+                "Slot2": ["s2", "s4"],
+                "Frac1": [0.1, 0.6],
+                "Frac2": [0.9, 0.4],
+                "Temp1": [10, 20],
+                "Temp2": [50, 60],
+                "Other": ["x", "y"],
+            },
+            [["Slot1", "Frac1", "Temp1"], ["Slot2", "Frac2", "Temp2"]],
+            {
+                "Slot1": ["s1", "s2", "s2", "s4"],
+                "Slot2": ["s2", "s4", "s1", "s2"],
+                "Frac1": [0.1, 0.6, 0.9, 0.4],
+                "Frac2": [0.9, 0.4, 0.1, 0.6],
+                "Temp1": [10, 20, 50, 60],
+                "Temp2": [50, 60, 10, 20],
+                "Other": ["x", "y", "x", "y"],
+            },
+            id="2inv_4dependent2each_1add",
+        ),
+    ],
+)
+def test_df_permutation_aug(data, col_groups, data_expected):
+    """Test permutation invariance data augmentation is done correctly."""
+    # Create all needed dataframes
+    df = pd.DataFrame(data)
+    df_augmented = df_apply_permutation_augmentation(df, col_groups)
+    df_expected = pd.DataFrame(data_expected)
+
+    # Determine equality ignoring row order
+    are_equal = (
+        pd.merge(left=df_augmented, right=df_expected, how="outer", indicator=True)[
+            "_merge"
+        ]
+        .eq("both")
+        .all()
+    )
+
+    assert (
+        are_equal
+    ), f"\norig:\n{df}\n\naugmented:\n{df_augmented}\n\nexpected:\n{df_expected}"
+
+
+@pytest.mark.parametrize(
+    ("col_groups", "msg"),
+    [
+        param([], "at least two column sequences", id="no_groups"),
+        param([["A"]], "at least two column sequences", id="just_one_group"),
+        param([["A"], ["B", "C"]], "the amount of columns in", id="different_lengths"),
+        param([[], []], "each column group has", id="empty_group"),
+    ],
+)
+def test_df_permutation_aug_invalid(col_groups, msg):
+    """Test correct errors for invalid permutation attempts."""
+    df = pd.DataFrame({"A": [1, 1], "B": [2, 2], "C": ["x", "y"]})
+    with pytest.raises(ValueError, match=msg):
+        df_apply_permutation_augmentation(df, col_groups)
+
+
+@pytest.mark.parametrize(
+    ("data", "causing", "affected", "data_expected"),
+    [
+        param(  # 1 causing val, 1 col affected (with 3 values)
+            {
+                "A": [0, 1],
+                "B": [3, 4],
+                "C": ["x", "y"],
+            },
+            ("A", [0]),
+            [("B", [3, 4, 5])],
+            {
+                "A": [0, 1, 0, 0],
+                "B": [3, 4, 4, 5],
+                "C": ["x", "y", "x", "x"],
+            },
+            id="1causing_1affected",
+        ),
+        param(  # 1 causing val, 2 cols affected (with 2 values each)
+            {
+                "A": [0, 1],
+                "B": [3, 4],
+                "C": ["x", "y"],
+            },
+            ("A", [0]),
```

Comment:
Not sure if in scope, thus only a question: The condition basically says "if A is 0, then it doesn't matter if B is 3 or 4 and it doesn't matter if C is "x" or "y", and B can be flipped independently of the flip in C", right? 

But do we also (need to) include the case of "Everything from above but now we also have groups D and E deactivated by A = 0, but the flipping should not be intermixed with the flipping of B and C; instead, the flipping needs to be handled group by group"?

Reply:
cannot follow please make an example

the current utils are aligned with the constraint, the constraint currently allows specifyingn "If A is ), then the values of B, C and D do not matter at all" which I guess just covers your first version? Any further functionality needs to be brought to the constraint first before talking about the util here

Reply:
Probably I can't follow my own example either 😄 

But you make a good point, perhaps arguing from the constraint side makes more sense. So looking at our example from the docs:
<img width="779" alt="image" src="https://github.com/emdgroup/baybe/assets/23265127/b053bf5e-c836-4be9-9c00-3cc97ac610c0">
Is it currently supported to say "Only if `Switch_1=on`, then `Solvent` and `Fraction` matter, but I **also** want that `X` and `Y` only matter if `Switch_1=off`"? If yes, how would that look like? Do I need to duplicate `Switch_1` in the parameters list? Or do I need to create a second constraint?

Reply:
yes thats simply a new entry in each `parameters`, `conditions` and `affected_parameters`

Reply:
Ok, makes sense. I think than that answers my previous question, since the augmentation utility will then simply process entry by entry in these collections, and this the individual entries will not interfere with each other in the augmentation, right? (that's what I meant above with `flipping should not be intermixed`)

Reply:
yep they shouldnt interfere (but its untested i think)

Reply:
Ok, perfect. But this cannot be tested on the utility level, would need to be tested on the constraint augmentation level, right? I think for now, we are good then. I think once we really integrate the augmentation, we should then add more tests, like this missing one, and perhaps also checks that actually validate the symmetry of the induced surrogate

Reply:
i think there should anyway be more augmentation tests reflecting such situations (@mhrmsn), but as you say these are beyond the cope of the utility and this PR